Some template backends like jinja doesn't have `engine` attribute

sodastsai opened the pull request:
So while doing offline compression, the command will crash.

diox commented:
That's weird : we do have tests covering jinja2 offline compression, and they pass. Can you add a test that would fail without that change ?

sodastsai commented:
I'm using [`django_jinja`](https://github.com/niwinz/django-jinja) as my jinja2 template loader with django 1.8. It delegates template loaders to the default django template loader, so it doesn't have an `engine` attributes.

sodastsai commented:
As [django's document](https://docs.djangoproject.com/en/1.8/topics/templates/#custom-backends), it doesn't require custom engines to implement an `engine` attribute.

brycedrennan commented:
I ran into this as well while using django_jinja

carltongibson commented:
Hmmm. There's obviously some difference to the way we're using Jinja and the way Django-Jinja does it... (I'm unfamiliar with that package so can't say.)

Do you still get the expected result with this change? (Are your jinja templates not missed?) How does Django-Jinja handle the loading?

We need some test to cover this. Can you provide one?

brycedrennan commented:
I was able to get offline compression working with django-jinja by using commit bcb2c43 along with the following setting:

```python
def COMPRESS_JINJA2_GET_ENVIRONMENT():
    from django.template import engines
    return engines['backend'].env
```

sodastsai commented:
Sorry for being busy last week.

As previous comment I posted, I found that django 1.8 doesn't require template engine backends to implement the `engine` attribute. Template backends like `django_jinja` and even the built-in jinja2 backend, which comes with django 1.8, don't implement it either. So in the first commit, I add a `has_attr` check while collecting `template_source_loaders` for django 1.8.

Since the 2 backends don't have the `engine` attributes, they could not provide any templates for django-compressor to perform offline compression. In my case, I use both `django_jinja` and original default django template backend (`django.template.backends.django.DjangoTemplates`). The default backend helps me to find all template files including jinja files, so the first fix works for me. 
The command I used to do offline compression is `python manage.py  compress --force --engine jinja2 -e .jinja`. So the default backend provides template loaders, finds templates, and filters them with `.jinja` extension. Things are going well.

But later, I created a new project for trying to answer @carltongibson's question, I found that, wow, if I only use `django_jinja` or only the built-in jinja2 backend from 1.8, the script won't work anymore. Yup, there's no one to provide templates for offline compilation.

Hence I added the second commit, it uses the `list_templates` method of Jinja2's environment (`jinja2.environment.Environment`) to find all templates provided by current environment. And then use the loader of current environment to expand them into file paths by using `get_source` method which is defined on `jinja2.loaders.BaseLoader`. By this way, I think the offline compilation would be work well with jinja2.

I'll add test later. I didn't use `tox` and write tests for django before, but I would try it.

sodastsai commented:
I just noticed that since `django-compressor` has its own template files (`compressor/templates/compressor/css_file.html` and etc.), it must be used with default django's template backend. (i.e., there's no such case that there would be only `django_jinja` backend or 1.8 built-in jinja2 backend)

carltongibson commented:
Slight aside, but I notice I disabled coveralls in 7c887546dec4990882b33c73d1c0cba2e406c9e9 (whilst working on a PR). @sodastsai Fancy adding that back in here so it can report? (My bad: I meant to replace it and forgot.)

Looking at it the change _seems_ ok — but it would be good to see an exact example of what fails without it to build a test around. Can you show that?

brkastner commented:
@sodastsai It looks like the templates will still work if you're only using jinja. The problem is compressor is currently using a Context instance instead of a dictionary for supplying the context to the template. This isn't the correct way to do it anymore so it would need to be modified to pass a dictionary to the template engine instead.  Compressor is currently skipping the jinja template backend because it doesn't support context instances in it's render_to_string method (which it shouldn't) and falling back to the built-in django template backend which still supports them.

sodastsai commented:
But if I only uses jinja backend which is provided by `django_jinja`, I got this error while performing offline compression with command `python manage.py compress --force --engine jinja2 -e .jinja --traceback`

```
(venv)sodas-macbook-pro:app-server sodas$ python manage.py compress --force --engine jinja2 -e .jinja --traceback
Found 'compress' tags in:
    /Users/sodas/work/app/app-server/web/templates/press_kit.jinja
    /Users/sodas/work/app/app-server/web/templates/courses.jinja
    /Users/sodas/work/app/app-server/web/templates/devices.jinja
    /Users/sodas/work/app/app-server/web/templates/index.jinja
    /Users/sodas/work/app/app-server/web/templates/_test.jinja
    /Users/sodas/work/app/app-server/web/templates/course_content.jinja
    /Users/sodas/work/app/app-server/web/templates/markdown.jinja
    /Users/sodas/work/app/app-server/web/templates/error.jinja
    /Users/sodas/work/app/app-server/web/templates/beta_expired.jinja
    /Users/sodas/work/app/app-server/web/templates/layouts/base.jinja
Compressing... Traceback (most recent call last):
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 244, in compress
    result = parser.render_node(template, context, node)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/offline/jinja2.py", line 105, in render_node
    return self._render_nodes(template, context, [node])
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/offline/jinja2.py", line 99, in _render_nodes
    return template.render(flat_context)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/jinja2/_compat.py", line 36, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 5, in <module>
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/contrib/jinja2ext.py", line 41, in _compress_forced
    return self._compress(kind, mode, caller, True)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/contrib/jinja2ext.py", line 52, in _compress
    return self.render_compressed(context, kind, mode, forced=forced)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/templatetags/compress.py", line 108, in render_compressed
    rendered_output = self.render_output(compressor, mode, forced=forced)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/templatetags/compress.py", line 121, in render_output
    return compressor.output(mode, forced=forced)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/css.py", line 51, in output
    ret.append(subnode.output(*args, **kwargs))
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/css.py", line 53, in output
    return super(CssCompressor, self).output(*args, **kwargs)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/base.py", line 298, in output
    return self.handle_output(mode, filtered_output, forced)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/base.py", line 306, in handle_output
    return output_func(mode, content, forced, basename)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/base.py", line 320, in output_file
    return self.render_output(mode, {"url": url})
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/base.py", line 345, in render_output
    return render_to_string(template_name, context_instance=final_context)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/template/loader.py", line 138, in render_to_string
    raise TemplateDoesNotExist(template_name)
django.template.base.TemplateDoesNotExist: compressor/css_file.html

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 11, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 338, in execute_from_command_line
    utility.execute()
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 330, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/core/management/base.py", line 390, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/core/management/base.py", line 441, in execute
    output = self.handle(*args, **options)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/django/core/management/base.py", line 657, in handle
    return self.handle_noargs(**options)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 293, in handle_noargs
    self.compress(sys.stdout, **options)
  File "/Users/sodas/work/app/app-server/venv/lib/python3.4/site-packages/compressor/management/commands/compress.py", line 247, in compress
    "%s" % (template.template_name, e))
django.core.management.base.CommandError: An error occured during rendering /Users/sodas/work/app/app-server/web/templates/press_kit.jinja: compressor/css_file.html
```

So that's why I think maybe I could not use only jinja backends. (Just got confused @@)

brkastner commented:
Are you using the built-in 1.8 jinja2 or django_jinja? Also, did you set COMPRESS_JINJA2_GET_ENVIRONMENT in your settings?

brkastner commented:
Do you  have COMPRESS_JINJA2_GET_ENVIRONMENT set in your settings?

sodastsai commented:
These are my `compressor` related settings

```
# `get_bool_environment` is a function which will get bool value from `os.environ` 
# and the second argument is the default value when there's no such env function
# Anyway, I'm sure that both `COMPRESS_ENABLED` and `COMPRESS_OFFLINE` are 
# `true` when running the offline compression
COMPRESS_ENABLED = get_bool_environment('COMPRESSOR_ENABLED', PRODUCTION)
COMPRESS_OFFLINE = get_bool_environment('COMPRESSOR_OFFLINE', RUNNING_ON_AWS)
COMPRESS_PRECOMPILERS = (
    ('text/less', 'lessc {infile} {outfile}' if not COMPRESS_ENABLED else 'lessc -x {infile} {outfile}'),
)
COMPRESS_JS_FILTERS = (
    'app.compressor.filters.uglify.UglifyFilter',  # My custom js filter used to compress js with uglify
)
COMPRESS_CSS_FILTERS = (
    'compressor.filters.css_default.CssAbsoluteFilter',
    'compressor.filters.cssmin.CSSMinFilter',
)
COMPRESS_OUTPUT_DIR = 'compressed'
COMPRESS_STORAGE = STATICFILES_STORAGE


def _compress_offline_context():
    from django.http import HttpRequest

    request = HttpRequest()
    request.LANGUAGE_CODE = LANGUAGE_CODE   # which is 'en'
    return {
        'request': request,
    }

COMPRESS_OFFLINE_CONTEXT = _compress_offline_context()


def COMPRESS_JINJA2_GET_ENVIRONMENT():
    from django.template import engines

    return engines['jinja2'].env
```

sodastsai commented:
oh, forgot to say that the error message I posted comes after applying my patch in this PR. If without this PR, the `compress` command just crashes and says it find an backend who doesn't have the `engine` attribute. (or it says it cannot find templates if you applies the first commit in this PR)

brkastner commented:
Okay so I did some digging and it looks like the TemplateDoesNotExist error isn't being thrown because the template doesn't exist, but instead because of the context issue. If you're willing to give me write access to your repository I can commit the change so it is included in your pull request. 

If not I'll just make a new pull request later today with the changes.

sodastsai commented:
@brkastner added :)

brkastner commented:
That should fix it! It was working for me without using offline compression with the version of django-compressor from pip so you might want to check if it does indeed fix the offline issue when combined with your fixes.

sodastsai commented:
@brkastner  :+1:

carltongibson commented:
OK — I feel really bad for asking for coveralls back — I shall see if there's a way of turning it down. (Sorry)

I'll have a more in depth play later on but, this looks good to me. It's all horribly gnarly — is there any test we can add that covers what's been _fixed_?

sodastsai commented:
The part @brkastner fixed is related to django 1.8's API deprecation. The signature of [`django.template.loader.render_to_string`](https://docs.djangoproject.com/en/1.8/topics/templates/#django.template.loader.render_to_string) has been changed. It doesn't accept [`django.template.Context`](https://docs.djangoproject.com/en/1.8/ref/templates/api/#django.template.Context) anymore. So, tests about this part is already covered by current one. (It's just change the API call to match django 1.8's change) (tests like `OfflineGenerationTestCaseWithContext`)
(btw, I suggest this part should be split into a new PR.)

---

And the part I fixed is that django's new Template API (which comes with 1.8) doesn't require each template backend has an `engine` attribute, so I added a [`hasattr` check while collecting `template_source_loaders`](https://github.com/django-compressor/django-compressor/pull/617/files#diff-a1dacfe6695942724d15beae6b957183R80)

so I added the default `jinja2` backend to the settings used in tests (see lines @ 7X in `compressor/tests/test_offline.py`). I think these would be enough for these fixes.

carltongibson commented:
@sodastsai Yes. The tests pass. Playing with it locally, nothing seems broken — but I'm not using Jinja. If nobody has any qualms I'd say merge it as is. 

@diox Do you have any issues here?

sodastsai commented:
Sure. I'm using it with mysite (https://tickleapp.com/) currently. At least it works for me :)

The site is running django 1.8 on python 3.4. The offline compression happens when the code is being deployed to the AWS EC2 instance. All the webpages on the site is also composed in jinja2 and assets like LESS files and javascript are compiled and compressed by django-compressor.

diox commented on compressor/tests/test_offline.py:
```diff
@@ -62,8 +62,44 @@ def setUp(self):
+            if django.VERSION < (1, 8):
+                override_settings["COMPRESS_JINJA2_GET_ENVIRONMENT"] = lambda: self._get_jinja2_env()
+            else:
+                override_settings['TEMPLATES'] += [
```
Why do you need to repeat all this ?

sodastsai replied:
Cuz in django 1.8, the `TEMPLATES` is a settings entry which must pass in a `dict`, and the `override_settings` must pass a root entry in. So that's why I have to repeat them again.

I mean, theses lines are used to override the `TEMPLATE_DIRS` populated at line 60, right? But I could not write such `override_settings['TEMPLATES'][0]['DIRS'] = override_settings['TEMPLATE_DIRS']`. The `override_settings` doesn't accept this.

diox replied:
deepcopy `settings['TEMPLATES']` before `self.settings()` is called, only changing `TEMPLATE_DIRS` inside your copy ?

sodastsai replied:
Huh, just watching the WWDC live stream ... though quite boring this year ...

anyway, tried to fix it by sodastsai/django-compressor@986ff0f

sodastsai commented:
Any updates?

La0 commented:
:+1: For this PR

diox commented on compressor/tests/test_offline.py:
```diff
@@ -62,7 +63,13 @@ def setUp(self):
 
-        if "jinja2" in self.engines:
+        if django.VERSION >= (1, 8):
+            tempaltes_settings = deepcopy(settings.TEMPLATES)
```
Typo: `tempaltes_settings`

sodastsai commented:
Fixed :)

coagulant commented:
I've tested this PR with latest stabel Django, django-jinja and django-compressor and the original problem with engine attribute missing is gone. :+1: 

However, I'm getting a lot of `TemplateSyntaxError` messages (`Invalid template`) during offline compression. This is because Jinja2Parser tries to parse non-jinja templates, loaded by another engine (in my case, `match_regex` option was not honoered). Probably, template loading mechanism in this management command is different from django 1.8 template loading mechanics.
I'll provide a demo of this problem later this week.

sodastsai commented:
@coagulant Any update on the problem you figured out?

sodastsai commented:
@coagulant I'm still waiting your feedback about this PR. :smile:

andreif commented:
Error messages are avoided in sodastsai/django-compressor#1

brkastner commented:
@andreif when you say error messages do you mean the ones @coagulant mentioned?

andreif commented:
@brkastner Yes, `Invalid template` ones

andreif commented:
Django admin is a good example of django templates in a mostly jinja2 project. Previously, they were passed to the jinja2 parser

diox commented:
We'd still like this to be merged in. What's the status ? Can you fix the conflicts ? Thanks.

sodastsai commented:
Sorry that I don't use django-compressor in my project now. (cuz we switched to Flask). And also this has been quite a long ago. So I re-organize the commit directly.

Could guys make sure these modification is enough to solve this issue?

diox commented:
Should be fixed in #715, thanks and apologies for the long wait.
